fix: add order transaction transition action pay

COMPRAWebDev opened the pull request:
<!--
Thank you for contributing to Shopware! Please fill out this description template to help us to process your pull request.

Please make sure to fulfill our contribution guideline (https://developer.shopware.com/docs/resources/guidelines/code/contribution?category=shopware-platform-dev-en/contribution).

Do your changes need to be mentioned in the documentation?
Please create a second pull request at https://github.com/shopware/docs
-->

### 1. Why is this change necessary?

When working with the `Shopware\Core\Checkout\Order\Aggregate\OrderTransaction\OrderTransactionStateHandler` there is currently no way to perform the `pay` transition action. 
But this action is required to execute one of the following `order_transaction` state transitions:
* `paid_partially` => `paid`
* `reminded` => `paid`

In addition to that, there is currently no constant in the class `Shopware\Core\System\StateMachine\Aggregation\StateMachineTransition\StateMachineTransitionActions` to address the required action `pay`.

### 2. What does this change do, exactly?

* Added constant `Shopware\Core\System\StateMachine\Aggregation\StateMachineTransition\StateMachineTransitionActions::ACTION_PAY` with value `pay`
* Added method `pay` to `Shopware\Core\Checkout\Order\Aggregate\OrderTransaction\OrderTransactionStateHandler` to handle the following `order_transaction` state transitions:
  * `paid_partially` => `paid`
  * `reminded` => `paid`
* Added test cases for the state transitions, mentioned above

### 3. Describe each step to reproduce the issue or behaviour.

Try to use the `OrderTransactionStateHandler` to process one of the `order_transaction` transitions, mentioned above. It's not possible with the current state of the code.

### 4. Please link to the relevant issues (if any).

[https://shopware.atlassian.net/browse/NEXT-40000](https://shopware.atlassian.net/browse/NEXT-40000)
[https://github.com/shopware/shopware/issues/5795](https://github.com/shopware/shopware/issues/5795)

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfill them.

COMPRAWebDev committed:
Add `order_transaction` transition action `pay`

COMPRAWebDev committed:
Removed empty line

COMPRAWebDev committed:
typo

COMPRAWebDev committed:
Added test cases for the state transitions, mentioned above

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/shopware/shopware?pullRequest=5803) <br/>All committers have signed the CLA.

COMPRAWebDev committed:
Merge branch 'trunk' into NEXT-40000/add-order-transaction-transition-action-pay

COMPRAWebDev committed:
Merge branch 'trunk' into NEXT-40000/add-order-transaction-transition-action-pay

COMPRAWebDev committed:
Merge branch 'trunk' into NEXT-40000/add-order-transaction-transition-action-pay

COMPRAWebDev committed:
Merge branch 'trunk' into NEXT-40000/add-order-transaction-transition-action-pay

COMPRAWebDev commented:
@mitelg Hi! This change is just adding some missing things without security breaks. Would be cool, if you guys from Shopware could check and simple merge it into the trunk.

lernhart commented:
Hey @COMPRAWebDev, thank you for contributing 🎉

It seems, that we gave the transaction actions very confusing names:

| Pre-State | Action | Post-State |
|--------|--------|--------|
| open | paid | paid | 
| paid_partially| pay | paid |
| reminded | pay | paid | 
| ... | ... | ... |

As you see, we are confusing actions "pay" and "paid". It should be "paid" for all action names. 

Would you mind to add a database migration for v6.7 to rename the following action names?
pay -> paid
pay_partially -> paid_partially

s. database extract
![image](https://github.com/user-attachments/assets/c4c175e7-f71b-496f-8268-7e6d7665105a)

Then you do not need the new action you proposed and cleaned up the whole thing in the process.

Here is a reference migration you could use as a starter: \Shopware\Core\Migration\V6_4\Migration1643878976AddCaptureRefundStateMachines

It uses the StateMachineMigrationTrait, which is a handy helper.

Thank you :)

COMPRAWebDev committed:
Added migration `Shopware\Core\Migration\V6_7\Migration1734604008RemoveDuplicateTransitionsForActionPay` to remove duplicate transitions for action_name = `pay`/`paid` resp. `pay_partially`/`paid_partially`

COMPRAWebDev committed:
Added migration `Shopware\Core\Migration\V6_7\Migration1734607798RenameTransitionActionPay` to rename transition actions `pay` and `pay_partially` to `paid` resp. `paid_partially`

COMPRAWebDev committed:
- Added new method `Shopware\Core\Checkout\Order\Aggregate\OrderTransaction\OrderTransactionStateHandler::paidPartially` to fulfill the new action name `paid_partially`
- Marked method `Shopware\Core\Checkout\Order\Aggregate\OrderTransaction\OrderTransactionStateHandler::payPartially` as deprecated, as the transition action `pay_partially` does not exist anymore

COMPRAWebDev committed:
Updated changelogs

COMPRAWebDev commented:
@lernhart please check the latest commits. We added two new migrations to 1.) remove duplicates for transitions with action "pay"/"paid" resp. "pay_partially"/"paid_partially" and 2.) rename the actions "pay" to "paid" and "pay_partially" to "paid_partially" + we added a new method "paidPartially" within the "OrderTransactionStateHandler" to fulfill the new general name of the action

COMPRAWebDev committed:
Removed superfluous ACTION_PAY

mitelg commented on src/Core/Checkout/Order/Aggregate/OrderTransaction/OrderTransactionStateHandler.php:
```diff
@@ -112,6 +112,8 @@ public function paid(string $transactionId, Context $context): void
     }
 
     /**
+     * @deprecated please use `paidPartially`, as the action `pay_partially` does not exist anymore.
```
this needs a proper deprecation format. you can find many examples if you search for `@deprecated`

COMPRAWebDev replied:
> this needs a proper deprecation format. you can find many examples if you search for `@deprecated`

Hi. We updated the deprecation annotation. Please check again

mitelg replied:
thanks, now @lernhart will take over for the rest :slightly_smiling_face:

COMPRAWebDev committed:
Updated deprecation annotation for method OrderTransactionStaeHandler::payPartially

lernhart commented:
@COMPRAWebDev 

Sorry, that this took so long, but we were in the midst of major release preparations for 6.7.
I've create a new PR [here](https://github.com/shopware/shopware/pull/7652), to be able to push on this. 

We've decided to move this to 6.8, as the deprecation for 6.7 would be too short of an upgrade window for extension developers.

Thank you for your contribution and your work so far :tada:
